### ISSUE TITLE
PublishReadyToRun fails on .NET6 / linux-musl-x64 / Docker

### ISSUE THREAD

**Issue by starburst997** [dotnet/runtime#62219]
### Description

I'm trying to build using `PublishReadyToRun` with `net6.0` in Docker (`mcr.microsoft.com/dotnet/sdk:6.0-alpine`) on `linux-musl-x64` and getting this error:

```
error NETSDK1095: Optimizing assemblies for performance is not supported for the selected target platform or architecture. Please verify you are using a supported runtime identifier, or set the PublishReadyToRun property to false.
```

My Docker file look like this:

```
FROM mcr.microsoft.com/dotnet/sdk:6.0-alpine AS build

COPY . .

RUN dotnet restore -r linux-musl-x64

RUN dotnet publish --framework net6.0 -c release -o /app -r linux-musl-x64 --self-contained true --no-restore /p:PublishTrimmed=true /p:PublishReadyToRun=true

FROM mcr.microsoft.com/dotnet/runtime-deps:6.0-alpine-amd64
WORKDIR /app
COPY --from=build /app ./

ENTRYPOINT ["./MyProgram"]
```

Using this exact Docker file with `net5.0` instead works without any issues.

### Reproduction Steps

Build a C# project using this Docker:

```
FROM mcr.microsoft.com/dotnet/sdk:6.0-alpine AS build

COPY . .

RUN dotnet restore -r linux-musl-x64

RUN dotnet publish --framework net6.0 -c release -o /app -r linux-musl-x64 --self-contained true --no-restore /p:PublishTrimmed=true /p:PublishReadyToRun=true
```

Notice that it fails with the error listed above.

### Expected behavior

Since it was building fine in `net5.0` I would expect `net6.0` to be able to build it as well.

### Actual behavior

It fails with:
```
error NETSDK1095: Optimizing assemblies for performance is not supported for the selected target platform or architecture. Please verify you are using a supported runtime identifier, or set the PublishReadyToRun property to false.
```

### Regression?

Yes this is a regression, it works perfectly in `net5.0`

### Known Workarounds

It seems like I can build just fine using windows (not Docker) instead but I haven't tried running the actual output but it did compiled without any error.

### Configuration

- Which version of .NET is the code running on?

`net6.0`

- What OS and version, and what distro if applicable?

`linux-musl-x64`

- What is the architecture (x64, x86, ARM, ARM64)?

`x64`

- Do you know whether it is specific to that configuration?

Seems like I can compile just fine from windows to `linux-musl-x64` but building using Docker using `mcr.microsoft.com/dotnet/sdk:6.0-alpine` or `mcr.microsoft.com/dotnet/sdk:6.0` fails.

**Comment by am11** [comment]
The issue is `--no-restore`, when we pass that flag, .NET 6 SDK skips downloading the platform package which is (implicitly) required by `dotnet-publish` (on first run). Delete `--no-restore` from dotnet publish line to see if it fixes the problem.

**Comment by starburst997** [comment]
Thank you very much! Yes that did work!

Still feels like a regression since it is impossible to specify to `dotnet restore` that I want it to restore with `PublishReadyToRun` so I can't have a distinct layer in my docker for restore (my docker file was trimmed down in the issue, I actually first copy only the `*.csproj`, do a `dotnet restore` then copy the rest of the files and finally do the `dotnet publish --no-restore`).

The error message is a bit misleading as well, makes it sounds like the platform cannot compile with PublishReadyToRun when in fact some files were not downloaded.

**Comment by am11** [comment]
Yup, this is a regression from .NET 5 SDK, and I agree that error message is not helpful. There are three ways to overcome it:

1. `dotnet publish` without `--no-restore` (this is workaround'ish)
2. persist `<PublishReadyToRun>true</PublishReadyToRun>` and `<PublishTrimmed>true</PublishTrimmed>` properties in csproj and continue to use `dotnet restore` and `dotnet publish --no-restore` (this is more sustainable, I think)
3. `dotnet restore -p:PublishReadyToRun=true -p:PublishTrimmed=true` then `donet publish --no-restore -p:PublishTrimmed=true -p:PublishReadyToRun=true ...` (i.e. do not change project properties between `restore` and `publish`)

**Comment by boukeversteegh** [comment]
Same problem, and can confirm that running `dotnet restore  -p:PublishReadyToRun=true ...` before the `dotnet --no-restore ...` fixes the issue. Thanks!

**Comment by wasabii** [comment]
Also hitting this. My situation is worse though. I'm invoking the Publish target multiple times during a build with different options. My build process outputs multiple platform specific outputs.

**Comment by boukeversteegh** [comment]
> Also hitting this. My situation is worse though. I'm invoking the Publish target multiple times during a build with different options. My build process outputs multiple platform specific outputs.

Could you share a sample of how you are currently using the restore command with multiple publish commands? I'm curious which parts are variable and which aren't, to see whether the above workaround applies.

**Comment by wasabii** [comment]
@boukeversteegh 

I have projects (msbuildproj files) that use the MSBuild task to invoke the Publish target of ProjectReferences, multiple times. Each time passing different RuntimeIdentifier values. For instance, my MSI setup building project has a set of ProjectReferences, that it loops over, and invokes Publish on: once for win7-x86, and another for win7-x64, and another for win7-arm. Generating three MSIs during the build, for each Windows platform.

I have another project that invokes the three multiple times targeting Linux, to generate x86, x64 and ARM output for ultimate inclusion in Docker containers.

I do this using MSBuild. Not in external scripts invoking 'dotnet'. Up until .NET 6, this worked fine, even allowing parallel builds of each of the different targets. The same project would be built in parallel for different platforms. MSBuild does that nearly automatically.

Now this is returning the error about optimization not being supported.

**Comment by boukeversteegh** [comment]
@wasabii So to build your solution with .NET 6, you would have to:

- call restore for each build, turning a _constant_ restore time into _**linear**_ , over number of build targets
- and either
  - run each publish command in sequence, turning what was a _constant_ publish time into _**exponential**_
  - copy the solution for each build target and publishing in parallel, adding another **linear** operation (over solution size)

I see how this is a major regression for solutions with many projects. Hopefully it will get some attention!

**Comment by wasabii** [comment]
Yeah. I can't do this. .NET 6 is just off limits for me for now.

**Comment by xPudda** [comment]
> Yup, this is a regression from .NET 5 SDK, and I agree that error message is not helpful. There are three ways to overcome it:
> 
> 2. persist `<PublishReadyToRun>true</PublishReadyToRun>` and `<PublishTrimmed>true</PublishTrimmed>` properties in csproj and continue to use `dotnet restore` and `dotnet publish --no-restore` (this is more sustainable, I think)

Solution number 2 works well ✅. Thank you @am11!

**Comment by wasabii** [comment]
Just a reminder that this is still a problem with NET 7 preview 3.

At this point I'm evaluating disabling ReadyToRun for my application.

**Comment by mangod9** [comment]
have you tried doing a restore with `<PublishReadyToRun>true</PublishReadyToRun>` enabled for one of the projects. Once the crossgen2 nuget is downloaded it should suffice for future projects. @AntonLapounov in case there are other ideas.

**Comment by ds38723** [comment]
https://learn.microsoft.com/en-us/dotnet/core/compatibility/sdk/6.0/publish-readytorun-requires-restore-change

**Comment by BDisp** [comment]
I also was facing this issue and found the cause was the `sdk` version of the `global.json` pointing to a non existing runtime version. So, if the version is set to `7.0.100` and the machine only have the `7.0.102` installed, then this issue will arise. The solution is change the sdk version to 7.0.102 or install the sdk version 7.0.100.

**Comment by Eternal21** [comment]
Hit the same error when trying to build a docker image.  Ended up having to disable `PublishReadyToRun` on project level:
`<PublishReadyToRun>false</PublishReadyToRun>`
